compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -4010,60 +4048,76 @@ public void visit(BLangWorkerSend workerSendNode) {
+        validateWorkerAnnAttachments(returnNode.expr, data);
+        data.notCompletedNormally = true;
     }
 
-    BType analyzeDef(BLangNode node, SymbolEnv env) {
-        return analyzeNode(node, env);
+    void analyzeDef(BLangNode node, AnalyzerData data) {
+        analyzeNode(node, data);
     }
 
-    BType analyzeStmt(BLangStatement stmtNode, SymbolEnv env) {
-        return analyzeNode(stmtNode, env);
+    void analyzeStmt(BLangStatement stmtNode, AnalyzerData data) {
+        analyzeNode(stmtNode, data);
     }
 
-    public BType analyzeNode(BLangNode node, SymbolEnv env) {
-        return analyzeNode(node, env, symTable.noType, null);
+    public void analyzeNode(BLangNode node, SymbolEnv env) {
+        AnalyzerData data = new AnalyzerData();
+        data.env = env;
+        analyzeNode(node, data);
+    }
+
+    public void analyzeNode(BLangNode node, SymbolEnv env, Stack<SymbolEnv> prevEnvs) {
+        AnalyzerData data = new AnalyzerData();
+        data.env = env;
+        data.prevEnvs = prevEnvs;
+        analyzeNode(node, data);
+    }
+
+    public void analyzeNode(BLangNode node,  AnalyzerData data) {
+        analyzeNode(node, data.env, symTable.noType, null, data);
     }
 
     @Override
-    public void visit(BLangContinue continueNode) {
-        this.notCompletedNormally = true;
+    public void visit(BLangContinue continueNode, AnalyzerData data) {
+        data.notCompletedNormally = true;
     }
 
     @Override
-    public void visit(BLangBreak breakNode) {
-        this.notCompletedNormally = true;
-        this.breakFound = true;
+    public void visit(BLangBreak breakNode, AnalyzerData data) {
+        data.notCompletedNormally = true;
+        data.breakFound = true;
     }
 
     @Override
-    public void visit(BLangPanic panicNode) {
-        this.typeChecker.checkExpr(panicNode.expr, env, symTable.errorType);
-        this.notCompletedNormally = true;
+    public void visit(BLangPanic panicNode, AnalyzerData data) {
+        this.typeChecker.checkExpr(panicNode.expr, data.env, symTable.errorType, data.prevEnvs);
+        data.notCompletedNormally = true;
     }
 
-    BType analyzeNode(BLangNode node, SymbolEnv env, BType expType, DiagnosticCode diagCode) {
-        this.prevEnvs.push(this.env);
-        BType preExpType = this.expType;
-        DiagnosticCode preDiagCode = this.diagCode;
+    void analyzeNode(BLangNode node, SymbolEnv env, BType expType, DiagnosticCode diagCode, AnalyzerData data) {
+        data.prevEnvs.push(data.env);
+        BType preExpType = data.expType;
+        DiagnosticCode preDiagCode = data.diagCode;
 
         // TODO Check the possibility of using a try/finally here
-        this.env = env;
-        this.expType = expType;
-        this.diagCode = diagCode;
-        node.accept(this);
-        this.env = this.prevEnvs.pop();
-        this.expType = preExpType;
-        this.diagCode = preDiagCode;
-
-        return resType;
+        if (data.newEnv != null) {
```

Comment:
As we discussed in the code review If we are going to remove the `newEnv` field then we need to add a new parameter(new function similar to `analyzeNode(BLangNode node, SymbolEnv newEnv, BType expType, DiagnosticCode diagCode, AnalyzerData data)`) to the function to pass this `newEnv`(Currently we pass it as a field in AnalyzerData). 
Are we going to pass it as a separate parameter or as a field in `AnalyzerData`? @hasithaa @MaryamZi @KavinduZoysa @lasinicl  WDYT?
Similar scenario discussed in https://github.com/ballerina-platform/ballerina-lang/pull/34909#discussion_r796662274

Reply:
Solved.